docs/src/docs/asciidoc/user/resilience.adoc

```diff
@@ -0,0 +1,189 @@
+
+=== Active and mirror setup
+
+It works the same as with a single server. Except that during a failover, the client will behaves like if the underlying
+server is having hiccups. Or is down if the failover takes too long.
+
+The new active server will notify the client when ready.
+
+=== Reconnection
+
+When a server goes down and back again. On the same URL. The client will silently reconnect to it. The ledger should allow
+to invalidate entries on the caching tier.
+
+== Loader writer
+
+Cache loader writers have a dedicated resilience strategy. So in case of failure, the resilience strategy will operate and
+the result will be picked up by the Cache resilience strategy.
+
+When using a loader writer, we tend to consider the following:
+
+* We expect the cache never return null. Because the loader will load any missing data. But to enforce that
+in case of server failure, it means the client should wait until the backend is back.
+* We expect that everything cached is saved. So in general an exception should be thrown if we fail to do so.
+
+The default resilience strategy is the following.
+
+* V load(K key): Retry infinitely
+* Map<K, V> loadAll(Iterable<? extends K> keys): Retry infinitely
+* void write(K key, V value): Exception
+* void writeAll(Iterable<? extends Map.Entry<? extends K, ? extends V>> entries): Exception
+* void delete(K key): Exception
+* void deleteAll(Iterable<? extends K> keys): Exception
+
+=== Write behind
+
+The resilience strategy is applied for write behind as well. So make sure it manages errors
+that are asynchronous.
+
+== Interruptions
+
+When waiting on a call to a store, an interruption should allow to get out. It will then probably rely on the resilience
+strategy (I'm not sure about that) or throw an exception right away.
+
+== Resilience Strategy
+
+The default resilience strategy (`RobustResilienceStrategy`) will
+
+* Return null on a read
+* Do nothing on a void mutation
+* Throw an exception if the result depends on the result of the mutation
+
+We can build your own strategy or used other provided ones.
+
+=== Alternative resilience strategies
+
+==== NOP
+
+This resilience strategy should provide a behavior similar to a caching that does not cache anything.
+I think this will be favorite default. At least without loader-writer.
+
+[cols=">s,1*", options="header"]
+|======================================================
+|Method         |Behavior
+|clear          |Do nothing
+|containsKey    |Return false
+|forEach        |empty list
+|get            |Return null
+|getAll         |Return all nulls
+|getAndPut      |Return null, nothing is put
+|getAndRemove   |Return null, nothing is removed
+|getAndReplace  |Return null, nothing is replaced
+|invoke         |Do nothing and return null
+|invokeAll      |Do nothing and return nulls
+|iterator       |Empty iterator
+|loadAll        |Do nothing
+|put            |Do nothing
+|putAll         |Do nothing
+|putIfAbsent    |Do nothing, return null
```

Comment:
This seems weird... the null return is at odds with the "do nothing", since a null return implies the method did something.

Reply:
It's the cache not caching. So `putIfAbsent` will consider the cache to be empty but won't cache anything. You can see it like a cache with entries expiring at insertion. Or a cache of size 0 with everything evicted immediately.

Reply:
I guess that makes sense.

Reply:
What is the difference? Returning null will cause the XA transaction to commit. If the user wants a rollback in case of cache failure, he better use a resilience strategy that matches his needs.